torch/csrc/jit/script/compiler.cpp

```diff
@@ -477,77 +477,74 @@ struct to_ir {
 
       // Add block outputs
       for (const auto& x : body_frame->captured_inputs) {
-        body_block->registerOutput(body_frame->getValueInThisFrame(stmt.range(), x));
-        n->addInput(outer_frame->getVar(x, stmt));
+        body_block->registerOutput(body_frame->getValueInThisFrame(range, x));
+        n->addInput(outer_frame->getVar(x, range));
         outer_frame->setVar(x, n->addOutput());
       }
 
     }
   }
 
+  void emitForRange(SourceRange range, const Ident& target, const List<Expr>& args, const List<Stmt>& body) {
+    // TODO: start, stop, step loop
+    if (args.size() != 1) {
+      throw ErrorReport(range)
+          << "range() expects one argument but got" << args.size();
+    }
+    emitLoopCommon(range, {args[0]}, {}, body, target);
+  }
+
   void emitFor(const For& stmt) {
     // For now, we only support range loops. e.g. for i in range(3): ...
-
     auto targets = stmt.targets();
     auto itrs = stmt.itrs();
     auto body = stmt.body();
 
-    // itrs must consist of a single Apply node
     if (stmt.itrs().size() != 1) {
       throw ErrorReport(stmt)
           << "List of iterables is not supported currently.";
     }
-    if (itrs[0].kind() != TK_APPLY) {
-      throw ErrorReport(stmt)
-          << "Non-range for loops are currently not supported.";
+    if (targets.size() != 1) {
+      throw ErrorReport(stmt) << "Iteration variable unpacking is not supported";
     }
 
-    Apply range_iterator = Apply(itrs[0]);
-    if (range_iterator.callee().kind() != TK_VAR) {
-      throw ErrorReport(stmt)
-          << "Non-range for loops are currently not supported.";
-    }
-
-    {
-      Var var = Var(range_iterator.callee());
-      if (var.name().name() != "range") {
-        throw ErrorReport(stmt)
-            << "Non-range for loops are currently not supported.";
+    // match range(<expr>) style loops
+    // itrs must consist of a single Apply node
+    if (itrs[0].kind() == TK_APPLY) {
+      Apply range_iterator = Apply(itrs[0]);
+      if (range_iterator.callee().kind() == TK_VAR) {
+        Var var = Var(range_iterator.callee());
+        if (var.name().name() == "range") {
+          return emitForRange(stmt.range(), targets[0], range_iterator.inputs(), body);
+        }
       }
     }
 
-    List<Expr> args = range_iterator.inputs();
-    // TODO: start, stop, step loop
-    if (args.size() != 1) {
-      throw ErrorReport(stmt)
-          << "range() expects one argument but got" << args.size();
-    }
-
-    auto target_list = List<Ident>(targets);
-    if (target_list.size() != 1) {
-      throw ErrorReport(stmt)
-          << "Iteration variable unpacking is not supported";
+    // it isn't a range(<expr>) loop, treat it as a sugared value that maybe can be
+    // unrolled
+    auto sv = emitSugaredExpr(itrs[0]);
+    auto instances = sv->unrolledFor(stmt.range(), method);
+    for(auto inst : instances) {
+      environment_stack->setSugaredVar(targets[0].name(), inst);
```

Comment:
This will leave the iterator var in the environment, which is inconsistent with other kinds of loops IIRC.